### PR TITLE
[Windows] Refactor ingestion to avoid long tasks

### PR DESCRIPTION
**Pull request by MatkovIvan**
Refined version of #640

### CODE REVIEW

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/NetworkStateIngestion.cs [hunk 1]
```diff
@@ -1,124 +1,102 @@
 ﻿using System;
 using System.Collections.Generic;
-using System.Threading;
-using System.Threading.Tasks;
 using Microsoft.AppCenter.Ingestion.Models;
 
 namespace Microsoft.AppCenter.Ingestion.Http
 {
-    public class NetworkStateIngestion : IngestionDecorator
+    internal sealed class NetworkStateIngestion : IngestionDecorator
```
Comment (guperrot) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/NetworkStateIngestion.cs]: why sealed?

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/NetworkStateIngestion.cs [hunk 2]
```diff
@@ -1,124 +1,102 @@
...
     {
-        private readonly HashSet<IServiceCall> _calls = new HashSet<IServiceCall>();
-        private readonly SemaphoreSlim _mutex = new SemaphoreSlim(1, 1);
+        private readonly HashSet<ServiceCall> _calls = new HashSet<ServiceCall>();
```
Comment (guperrot) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/NetworkStateIngestion.cs]: does assigning to ISet work here?

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/RetryableIngestion.cs [hunk 3]
```diff
@@ -1,80 +1,152 @@
...
-            };
-
-        private readonly Func<Task>[] _retryIntervals;
+        {
+            TimeSpan.FromSeconds(10),
+            TimeSpan.FromSeconds(30),
+            TimeSpan.FromMinutes(20)
+        };
+        
+        private readonly IDictionary<ServiceCall, Timer> _calls = new Dictionary<ServiceCall, Timer>();
```
Comment (guperrot) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/RetryableIngestion.cs]: add a blank line

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/RetryableIngestion.cs [hunk 4]
```diff
@@ -1,80 +1,152 @@
...
 using System.Collections.Generic;
+using System.Linq;
 using System.Threading;
-using System.Threading.Tasks;
 using Microsoft.AppCenter.Ingestion.Models;
 
 namespace Microsoft.AppCenter.Ingestion.Http
 {
-    public class RetryableIngestion : IngestionDecorator
+    internal sealed class RetryableIngestion : IngestionDecorator
```
Comment (guperrot) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/RetryableIngestion.cs]: why sealed?

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/RetryableIngestion.cs [hunk 5]
```diff
@@ -1,80 +1,152 @@
...
-            var randomInt = Random.Next(max);
-            RandomMutex.Release();
-            return randomInt;
+            CancelAllCalls();
+            base.Dispose();
+        }
+
+        private void CancelAllCalls()
+        {
+            List<ServiceCall> calls;
```
Comment (guperrot) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/RetryableIngestion.cs]: IList?

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/ServiceCall.cs [hunk 6]
```diff
@@ -1,54 +1,156 @@
...
 
-namespace Microsoft.AppCenter.Ingestion.Http
+namespace Microsoft.AppCenter.Ingestion
 {
-    public abstract class ServiceCall : IServiceCall
+    /// <inheritdoc />
+    internal class ServiceCall : IServiceCall
     {
-        private CancellationTokenSource _tokenSource = new CancellationTokenSource();
+        private readonly CancellationTokenSource _tokenSource = new CancellationTokenSource();
```
Comment (guperrot) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/ServiceCall.cs]: blank line between fields

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/ServiceCall.cs [hunk 7]
```diff
@@ -1,54 +1,156 @@
...
+        public Exception Exception { get; private set; }
 
         public CancellationToken CancellationToken => _tokenSource.Token;
 
-        protected ServiceCall(IIngestion ingestion, IList<Log> logs, string appSecret, Guid installId)
+        public string AppSecret { get; }
+        public Guid InstallId { get; }
+        public IList<Log> Logs { get; }
+        
+        public ServiceCall()
```
Comment (guperrot) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/ServiceCall.cs]: use case for this empty one?

File: Tests/Microsoft.AppCenter.Test.Windows/Channel/ChannelTest.cs [hunk 8]
```diff
@@ -367,15 +367,14 @@ public void IngestionNotClosedOnRecoverableHttpError()
...
-            _mockIngestion.TaskError = new NonRecoverableIngestionException();
+            MakeIngestionCallsFail(new NonRecoverableIngestionException());
             _channel.EnqueueAsync(new TestLog()).RunNotAsync();
 
             // wait up to 20 seconds for suspend to finish
             bool disabled = WaitForChannelDisable(TimeSpan.FromSeconds(20));
             Assert.IsTrue(disabled);
-
-            Assert.IsTrue(_mockIngestion.IsClosed);
+
```
Comment (guperrot) [inline Tests/Microsoft.AppCenter.Test.Windows/Channel/ChannelTest.cs]: adding a tab by accident here

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/ServiceCall.cs (hunk 6 above)
Comment (MatkovIvan) [reply SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/ServiceCall.cs]: Do we use this convention here?
https://github.com/Microsoft/AppCenter-SDK-DotNet/blob/1.5.0/SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/AppCenter.cs#L20-L43

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/ServiceCall.cs (hunk 7 above)
Comment (MatkovIvan) [reply SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/ServiceCall.cs]: For tests

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/ServiceCall.cs (hunk 6 above)
Comment (guperrot) [reply SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/ServiceCall.cs]: It look wrong there too.

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/NetworkStateIngestion.cs (hunk 1 above)
Comment (MatkovIvan) [reply SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/NetworkStateIngestion.cs]: This is a decorator and other classes from should not be inherited from it. This is just guard.

**Comment by MatkovIvan** [comment]
@achocron We shouldn't do that

**Pull request merged**